### ISSUE TITLE
[QUESTION] AKS workload identity integration - how to put access token

### ISSUE THREAD

**Issue by gennadiy-dubina** [microsoft/mssql-jdbc#2058]
## Question
We are switching to using managed identity to access SQL. 
Our service is spring boot 3 app and it's hosted in AKS.
So we have configured workload identity and user managed identity in azure. Managed idenityt is added as sql user according to tutorial.
But in our setup we don't want to use any sidecar container to handle `ActiveDirectoryMSI`.

I have extended `HikariDataSource` and overrid `getPassword` to retrive access token via `ManagedIdentityCredential`
Now i can get token but we don't know how to put it to driver correctly
By default hikari puts it as `password` to driver properties and it does not work.
I have changed this behaviour and put token as `accessToken` to driver properties. Also it requires to remove `Authentication=ActiveDirectoryMSI` from jdbc url to avoid driver embeded handler
```
@Override
public Connection getConnection(final String username, final String password) throws SQLException {
	final var cloned = (Properties) driverProperties.clone();
	cloned.put("accessToken", password);
	return driver.connect(jdbcUrl, cloned);
}
````
but i'm not sure it will work correctly. Could you advise?
now i'm getting errro `Login failed for user '<token-identified principal>'. Incorrect or invalid token.`
Also what scope should i use to retrive access token? now i use `https://database.windows.net/.default`

P.S. looks like you have already extended driver to provide token but i can not find any release with these changes - https://github.com/microsoft/mssql-jdbc/pull/1940 

Thanks,

**Comment by gennadiy-dubina** [comment]
seems my approach works, 
i just need to have single scope in request "https://database.windows.net/.default"
so token will `"aud": "https://database.windows.net",`

```
new TokenRequestContext().addScopes("https://database.windows.net/.default")
```

btw when do you plan to release v12 of driver?

Thanks,

**Comment by lilgreenbird** [comment]
hi @gennadiy-dubina 

We are actually due to be releasing 12.2.0 tomorrow.

**Comment by lilgreenbird** [comment]
@gennadiy-dubina 

fyi we just released [12.2.0](https://github.com/microsoft/mssql-jdbc/releases)

**Comment by gennadiy-dubina** [comment]
@lilgreenbird i have reopened ticket because i have got issue with using `SQLServerAccessTokenCallback`
I have found we can not put it to `SQLServerDriver` via `suppliedProperties` (key: `accessTokenCallback`), it will be filtered out.

Seems we can use it with `SQLServerDataSource` only.
But it can not be used together with jdbc url anyway.
If i use `setAccessTokenCallback` with `setURL` it does not wokr because if  `dataSourceURL` is provided `SQLServerDataSource.getConnectionInternal` uses the same logic as `SQLServerDriver` and filters our it again.

**My code:**
```
@Bean
HikariDataSource dataSource(DataSourceProperties properties, AzureManagedIdentityAuth azureManagedIdentityAuth) {
    var msSqlDataSource = new SQLServerDataSource();
    msSqlDataSource.setURL(properties.getUrl());
    msSqlDataSource.setAccessTokenCallback(azureManagedIdentityAuth);
```

**_SQLServerDataSource.getConnectionInternal_ code** 

```
....
userSuppliedProps = connectionProps;
....
// Merge in URL properties into userSuppliedProps if URL is set.
if (null != dataSourceURL) {
    Properties urlProps = Util.parseUrl(dataSourceURL, dsLogger);
    // null returned properties means that the passed in URL is not
    // supported.
    if (null == urlProps)
        SQLServerException.makeFromDriverError(null, null,
                SQLServerException.getErrString("R_errorConnectionString"), null, true);
    // Manually merge URL props and user supplied props.
    mergedProps = SQLServerDriver.mergeURLAndSuppliedProperties(urlProps, userSuppliedProps);
} else {
    mergedProps = userSuppliedProps;
}
```

So you will drop `accessTocketCallback` property from `userSuppliedProps` (it's `connectionProps`) with line
```
mergedProps = SQLServerDriver.mergeURLAndSuppliedProperties(urlProps, userSuppliedProps);
````
https://github.com/microsoft/mssql-jdbc/blob/main/src/main/java/com/microsoft/sqlserver/jdbc/SQLServerDataSource.java#L1456 


So why we can not use jdbc connection string with `setAccessTokenCallback` ?

Thanks,

**Comment by tkyc** [comment]
Hi, that is a known limitation. `setAccessTokenCallback` will only work with datasources as that was the original feature request. Connection string support will be coming out in the next preview release.

**Comment by gennadiy-dubina** [comment]
@tkyc it does not work with datasource too, my example
```
var msSqlDataSource = new SQLServerDataSource();
msSqlDataSource.setURL(properties.getUrl()); //jdbc url
msSqlDataSource.setAccessTokenCallback(azureManagedIdentityAuth);
```
maybe "datasource" means - specify all parmeters manually

ok. will wait next release 

thx

**Comment by gennadiy-dubina** [comment]
one more question about token expiration time: if we use accessTokenCallback and token is expired does SqlConnection will retrive new token from callback automatically or we need to care about manually?

**Comment by tkyc** [comment]
> maybe "datasource" means - specify all parmeters manually

@gennadiy-dubina, that is correct. You must specify the datasource params manually through the setters for the datasource and not through the URL. A working example of it can be found [here](https://github.com/microsoft/mssql-jdbc/blob/main/src/samples/azureactivedirectoryauthentication/src/main/java/AzureActiveDirectoryAccessTokenCallback.java).

> if we use accessTokenCallback and token is expired does SqlConnection will retrive new token from callback automatically or we need to care about it manually?

The driver will use the callback to renew the token when expiry occurs. You do not need to manually manage token renewal. This feature was meant to replace the old `setAccessToken` for datasources, as that doesn't renew the token.

**Comment by gennadiy-dubina** [comment]
@tkyc thank you
seems my question about token revewal does not make any sense.
looks like it's used in the beginig only when we open sql connection. so after that token could expire but opened sql connection will work without any issue. we just need probide fresh token when we open new sql connection.
It means we can keep sql connection in the pool as long as we want.

**Comment by tkyc** [comment]
> looks like it's used in the beginig only when we open sql connection. so after that token could expire but opened sql connection will work without any issue. we just need probide fresh token when we open new sql connection.
It means we can keep sql connection in the pool as long as we want.

> is it possible that connection will be closed automaticaly by server or returns "invalid credentials" error for sql command when token is expired?

@gennadiy-dubina With the callback, it shouldn't be possible for a token expiry error to occur when making a query. When I said the driver renews the token it wasn't technically true. What exactly happens is when a token expiry occurs, the driver will use the callback to fetch a new token. So, you don't need to use refresh tokens or manually renew the token in any way. Any connections in the pool with an expired token will be closed and replaced with new connections with non-expired tokens. So, the driver will manage this all for you.

**Comment by tkyc** [comment]
@gennadiy-dubina I've attached a driver jar with the connection string changes for the callback to this comment. The new connection string property/value pair is `accessTokenCallbackClass=<fully-qualified-class-name>` (to get the fully qualified class name you can do `<your-class>.class.getName()`). And I've also fixed the `setURL` for the callback as well. Please try it out if possible and let me know of anything that can be improved on. This new change will be available in the upcoming 12.3.0 preview. Thanks.

[driver.zip](https://github.com/microsoft/mssql-jdbc/files/10702653/driver.zip)
